darts/utils/data/simple_inference_dataset.py

```diff
@@ -11,17 +11,88 @@
+        are provided, otherwise a value of `None` will be emitted.
+        Both `tgt_past` and `cov_past` will be `input_chunk_length` long.
+        If the model is required to produce the forecast over multiple iterations, i.e. if
+        `n > self.output_chunk_length`, and if covariates were provided, then `cov_future` will
+        be equal to the future covariates up to `n - self.output_chunk_length` time steps into the future.
+
+        Parameters
+        ----------
+        series
+            The target series that are to be predicted into the future.
+        covariates
+            Optionally, the corresponding covariates that are used for predictions. This argument is required
+            if the model was trained with covariates.
+        n
+            The number of time steps after the end of the training time series for which to produce predictions.
+        input_chunk_length
+            The length of the time series the model takes as input.
+        output_chunk_length
+            The length of the model predictions after one call to its `forward` function.
+    """
+
         super().__init__()
         self.series = [series] if isinstance(series, TimeSeries) else series
         self.covariates = [covariates] if isinstance(covariates, TimeSeries) else covariates
-        raise_if_not(covariates is None or len(series) == len(covariates),
+        self.n = n
+        self.input_chunk_length = input_chunk_length
+        self.output_chunk_length = output_chunk_length
+
+        raise_if_not((covariates is None or len(series) == len(covariates)),
                      'The number of target series must be equal to the number of covariates.')
 
     def __len__(self):
         return len(self.series)
 
-    def __getitem__(self, idx: int) -> Tuple[TimeSeries, Optional[TimeSeries]]:
-        return (self.series[idx], None) if self.covariates is None else (self.series[idx], self.covariates[idx])
+    def __getitem__(self, idx: int) -> Tuple[TimeSeries, Optional[TimeSeries], Optional[TimeSeries]]:
+
+        target_series = self.series[idx]
+        covariate_series = None if self.covariates is None else self.covariates[idx]
+
+        raise_if_not(len(target_series) >= self.input_chunk_length,
+                     'All input series must have length >= `input_chunk_length` ({}).'.format(
+                     self.input_chunk_length))
+
+        tgt_past = target_series[-self.input_chunk_length:]
+
+        cov_past = cov_future = None
+        if covariate_series is not None:
+
+            # get first timestamp that lies in the future of target series
+            first_pred_time = target_series.end_time() + target_series.freq()
+
+            # isolate past covariates and add them to array
+            if covariate_series.end_time() >= first_pred_time:
+                cov_past = covariate_series.drop_after(first_pred_time)
```

Comment:
All of these commands (drop_after, indexing with [], etc...) create new `TimeSeries` objects, which means behind the scenes a lot of object creation, and data copying (remember the whole data is copied over whenever a series is created). I suspect the performance might be much better if we make __getitem__ return np.ndarray's instead, and then access the values using `series.values(copy=False)`, very much like we do within the training datasets. It's (slightly) uglier, but it has the big advantage of manipulating only Numpy views of the underlying arrays, which I think will be much faster in some cases.
WDYT?

Reply:
Yeah I see what you mean. However, the problem is that we need the time indices at the time of prediction. I think that was the initial motivation between making `SimpleInferenceDataset` return `TimeSeries` instances instead of numpy arrays. Strictly speaking, we only need the time indices of the target series. So maybe we could have a mixed return type, only creating `TimeSeries` objects for the target series, but not for the covariates to at least save some resources? Or I guess we could separately emit the timestamp that indicates the start of the prediction and have the target and covariate values be numpy arrays?

Reply:
You're right, we do need the time indices.
I would then propose that we leave it like that for now, and add a task in our backlog to revisit this at some later point of time. We could then do like you propose (emit only the start of the prediction Timestamp).